File: app/src/main/java/com/github/panda_io/micro_panda/ast/expression/Expression.java
package com.github.panda_io.micro_panda.ast.expression;

import com.github.panda_io.micro_panda.ast.Node;
import com.github.panda_io.micro_panda.ast.type.Type;
import com.github.panda_io.micro_panda.ast.Context;

public abstract class Expression extends Node {
    boolean constant;
    Type type;

    public Type getType() {
        return this.type;
    }

	public boolean isConstant() {
        return this.constant;
    }

	public abstract void validate(Context context, Type expected);
}

File: MicroPanda/AST/Expression/Expression.cs
namespace MicroPanda.AST.Expression;

using Node;
using Type;

internal abstract class Expression : Node
{
    protected bool _const;
    protected Type? _type;
    internal abstract void Validate(Context context, Type? expected);
    internal bool Const => _const;
    internal Type? Type => _type;
}